File: backend/src/worker/monitor.rs
use std::sync::Arc;
use crate::{app::AppState, db::queries};

/// Recover jobs that were in progress when the server stopped.
pub async fn recover_stale_jobs(state: &Arc<AppState>) {
    let processing_statuses = [
        "processing", "downloading", "checkpoint_download",
        "transcoding", "checkpoint_transcode", "uploading",
    ];

    let stale = match queries::list_jobs_by_statuses(&state.db, &processing_statuses).await {
        Ok(jobs) => jobs,
        Err(e) => {
            tracing::error!("Failed to query stale jobs: {}", e);
            return;
        }
    };

    let count = stale.len();

    for job in &stale {
        tracing::warn!("Recovering stale job {} (status: {})", job.id, job.status);
        queries::update_job_failed(
            &state.db, &job.id,
            "Server restarted — job interrupted, please retry",
        ).await.ok();

        queries::insert_job_event(
            &state.db, &job.id, None, "error",
            "Server restarted while job was in progress", None,
        ).await.ok();
    }

    if count == 0 {
        tracing::info!("No stale jobs to recover");
    } else {
        tracing::info!("Recovered {} stale job(s)", count);
    }
}

File: backend-bun/src/worker/monitor.ts
import type { DrizzleDB } from "../db/index";
import type { EventBus } from "../api/events";
import {
  listJobsByStatuses,
  updateJobFailed,
  insertJobEvent,
} from "../db/queries";

const ACTIVE_STATUSES = [
  "processing",
  "downloading",
  "checkpoint_download",
  "transcoding",
  "checkpoint_transcode",
  "uploading",
];

/**
 * Recover jobs left in active states after a server restart.
 * Sets each to failed with a human-readable message so the user can retry.
 */
export function recoverStaleJobs(db: DrizzleDB, _eventBus?: EventBus): void {
  const staleJobs = listJobsByStatuses(db, ACTIVE_STATUSES);
  for (const job of staleJobs) {
    updateJobFailed(db, job.id, "Server restarted — job interrupted, please retry");
    insertJobEvent(
      db,
      job.id,
      null,
      "error",
      "Server restarted while job was in progress",
      null,
    );
  }
  console.log(`[monitor] Recovered ${staleJobs.length} stale job(s)`);
}
